src/pw/pw_fpga.F

```diff
@@ -0,0 +1,279 @@
+      FUNCTION pw_fpga_check_bitstream(n) RESULT(res) &
+         BIND(C, name="pw_fpga_check_bitstream_")
+         IMPORT
+         INTEGER(KIND=C_INT), DIMENSION(*), &
+            INTENT(IN)                            :: n
+         LOGICAL(KIND=C_BOOL)                     :: res
+
+      END FUNCTION pw_fpga_check_bitstream
+
+   END INTERFACE
+
+   INTERFACE
+! **************************************************************************************************
+!> \brief single precision FFT3d using FPGA
+!> \param data_path_len - length of the path to the data directory
+!> \param data_path - path to the data directory
+!> \param dir - direction of FFT3d
+!> \param npts - dimensions of FFT3d
+!> \param single precision c_in...
+! **************************************************************************************************
+      SUBROUTINE pw_fpga_fft3d_sp(data_path_len, data_path, dir, n, c_in_sp) &
+         BIND(C, name="pw_fpga_fft3d_sp_")
+         IMPORT
+         INTEGER(KIND=C_INT), VALUE, INTENT(IN)  :: data_path_len
+         CHARACTER(len=1, kind=C_CHAR), TARGET    :: data_path(data_path_len)
+         INTEGER(KIND=C_INT), VALUE, INTENT(IN)   :: dir
+         INTEGER(KIND=C_INT), DIMENSION(*), &
+            INTENT(IN)                            :: n
+         COMPLEX(KIND=C_FLOAT_COMPLEX)           :: c_in_sp(n(1), n(2), n(3))
+      END SUBROUTINE pw_fpga_fft3d_sp
+   END INTERFACE
+
+   INTERFACE
+! **************************************************************************************************
+!> \brief double precision FFT3d using FPGA
+!> \param data_path_len - length of the path to the data directory
+!> \param data_path - path to the data directory
+!> \param dir - direction of FFT3d
+!> \param npts - dimensions of FFT3d
+!> \param double precision c_in...
+! **************************************************************************************************
+      SUBROUTINE pw_fpga_fft3d_dp(data_path_len, data_path, dir, n, c_in_dp) &
+         BIND(C, name="pw_fpga_fft3d_dp_")
+         IMPORT
+         INTEGER(KIND=C_INT), VALUE, INTENT(IN)  :: data_path_len
+         CHARACTER(len=1, kind=C_CHAR), TARGET    :: data_path(data_path_len)
+         INTEGER(KIND=C_INT), VALUE, INTENT(IN)   :: dir
+         INTEGER(KIND=C_INT), DIMENSION(*), &
+            INTENT(IN)                            :: n
+         COMPLEX(KIND=C_DOUBLE_COMPLEX)           :: c_in_dp(n(1), n(2), n(3))
+      END SUBROUTINE pw_fpga_fft3d_dp
+   END INTERFACE
+
+CONTAINS
+
+! **************************************************************************************************
+!> \brief Allocates resources on the fpga device
+! **************************************************************************************************
+   SUBROUTINE pw_fpga_init()
+#if defined (__PW_FPGA)
+      INTEGER :: stat
+
+#if defined (__PW_CUDA)
+#error "CUDA and FPGA cannot be configured concurrently!"
```

Comment:
fyi the "#error@ directive may not be defined for all Fortran pre-processors

Reply:
Added CPABORT statements after the error directive so that they fail on runtime in case the preprocessor doesn't catch this.